### ISSUE TITLE
Add HTTP Basic Auth

### ISSUE THREAD

**Issue by jaideep2** [jaideep2/apinaio#5]
Implement this
> Here's how you add HTTP Basic Auth with `caddy` as a reverse proxy to `localhost:11434`, and also handle HTTPS automatically:
> 
> 0. Install caddy
>    ```sh
>    # Mac, Linux
>    curl https://webi.sh/caddy | sh
> 
>    # Windows
>    curl.exe https://webi.ms/caddy | powershell
>    ```
> 1. Put your password (which could be an API Token) in a `password.txt`
> 2. Digest the password
>    ```sh
>    caddy hash-password < ./password.txt
>    ```
> 3. Put the username and digest in an ENV file \
>    `caddy.env`:
>    ```
>    BASIC_AUTH_USER='apitoken'
>    BASIC_USER_AUTH='$2a$14$sI1j0RbhzKHMZ4cHU8otHOkB3Dgl9egF2D.CXB6C0/Qk5dtaMHS/u'
>    ```
> 4. Create a `Caddyfile` with basic auth using the ENVs
>    ```Caddyfile
>    api.example.com {
>        handle /* {
>            basicauth {
>                {env.BASIC_AUTH_USERNAME} {env.BASIC_AUTH_DIGEST}
>            }
>            reverse_proxy localhost:11434
>        }
>    }
>    ```
> 5. Run caddy
>    ```sh
>    caddy run --config ./Caddyfile --envfile ./caddy.env
>    ```
> 
> And if you want to run it as a system service, or _without_ HTTPS or need other details, I've got a bunch of snippets up at <https://webinstall.dev/caddy>.
> 

_Originally posted by @coolaj86 in https://github.com/jmorganca/ollama/issues/849#issuecomment-1773697189_

**Comment by followkim** [comment]
Sure sure if this is still being watched, but I am getting an error deploying this:

kim@kim:~$ caddy run --config ./CaddyFile --envfile ./caddy.env
2025/02/12 03:33:24.230	INFO	using config from file	{"file": "./CaddyFile"}
2025/02/12 03:33:24.231	INFO	adapted config to JSON	{"adapter": "caddyfile"}
2025/02/12 03:33:24.231	WARN	Caddyfile input is not formatted; run 'caddy fmt --overwrite' to fix inconsistencies	{"adapter": "caddyfile", "file": "./CaddyFile", "line": 2}
2025/02/12 03:33:24.232	INFO	admin	admin endpoint started	{"address": "localhost:2019", "enforce_origin": false, "origins": ["//localhost:2019", "//[::1]:2019", "//127.0.0.1:2019"]}
2025/02/12 03:33:24.232	INFO	http.auto_https	server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS	{"server_name": "srv0", "https_port": 443}
2025/02/12 03:33:24.232	INFO	http.auto_https	enabling automatic HTTP->HTTPS redirects	{"server_name": "srv0"}
2025/02/12 03:33:24.232	INFO	tls.cache.maintenance	started background certificate maintenance	{"cache": "0xc00050ef80"}
2025/02/12 03:33:24.232	INFO	tls.cache.maintenance	stopped background certificate maintenance	{"cache": "0xc00050ef80"}
Error: loading initial config: loading new config: loading http app module: provision http: server srv0: setting up route handlers: route 0: loading handler modules: position 0: loading module 'subroute': provision http.handlers.subroute: setting up subroutes: route 0: loading handler modules: position 0: loading module 'subroute': provision http.handlers.subroute: setting up subroutes: route 0: loading handler modules: position 0: loading module 'authentication': provision http.handlers.authentication: loading authentication providers: module name 'http_basic': provision http.authentication.providers.http_basic: account 0: username and password are required

**Comment by Ccocconut** [comment]
No, you implement it, it is out of scope of ollama. Use Caddy, nginx, whatever else you want. Do not introduce more complexity and bugs into ollama for doing what is not even its job.
